--- orig/py2app/bootstrap/argv_emulation.py	2017-06-12 17:21:32.000000000 +0100
+++ patch/py2app/bootstrap/argv_emulation.py	2017-06-12 17:22:44.000000000 +0100
@@ -93,7 +93,7 @@
     return carbon


-def _run_argvemulator(timeout=60):
+def _run_argvemulator(timeout=5):

     # Configure ctypes
     carbon = _ctypes_setup()
@@ -118,6 +118,7 @@
     FALSE = b'\0'
     TRUE = b'\1'
     eventLoopTimedOutErr = -9875
+    eventParameterNotFoundErr = -9870

     kEventClassAppleEvent, = struct.unpack('>i', b'eppc')
     kEventAppleEvent = 1
@@ -254,7 +255,7 @@

         sts = carbon.ReceiveNextEvent(
             1, ctypes.byref(eventType),
-            start + timeout[0] - now, TRUE, ctypes.byref(event))
+            start + timeout[0] - now, FALSE, ctypes.byref(event))

         if sts == eventLoopTimedOutErr:
             break
@@ -264,10 +265,17 @@
             break

         sts = carbon.AEProcessEvent(event)
-        if sts != 0:
+
+        # No events
+        if sts == eventParameterNotFoundErr:
+            break
+
+        elif sts != 0:
             print("argvemulator warning: processing events failed")
             break

+        now = time.time()
+
     carbon.AERemoveEventHandler(
         kCoreEventClass, kAEOpenApplication, open_app_handler, FALSE)
     carbon.AERemoveEventHandler(
